zebrad/src/components/mempool/storage/eviction_list.rs

```diff
@@ -0,0 +1,140 @@
+            max_size,
+            eviction_memory_time,
+        }
+    }
+
+    /// Inserts a TXID in the list, keeping track of the time it was inserted.
+    ///
+    /// All entries older than [`EvictionList::eviction_memory_time`] will be removed.
+    ///
+    /// # Panics
+    ///
+    /// If the TXID is already in the list.
+    ///
+    pub fn insert(&mut self, key: transaction::Hash) {
+        // From https://zips.z.cash/zip-0401#specification:
+        // > Nodes SHOULD remove transactions from RecentlyEvicted that were evicted more than
+        // > mempoolevictionmemoryminutes minutes ago. This MAY be done periodically,
+        // > and/or just before RecentlyEvicted is accessed when receiving a transaction.
+        self.prune_old();
+        // > Add the txid and the current time to RecentlyEvicted, dropping the oldest entry
+        // > in RecentlyEvicted if necessary to keep it to at most eviction_memory_entries entries.
+        if self.len() >= self.max_size {
+            self.pop_front();
+        }
+        let value = Instant::now();
+        let old_value = self.unique_entries.insert(key, value);
+        // It should be impossible for an already-evicted transaction to be evicted
+        // again since transactions are not added to the mempool if they are evicted,
+        // and the mempool doesn't allow inserting two transactions with the same
+        // hash (they would conflict).
+        assert!(
+            old_value.is_none(),
+            "an already-evicted transaction should not be evicted again"
+        );
+        self.ordered_entries.push_back(key)
+    }
+
+    /// Checks if the given TXID is in the list.
+    pub fn contains_key(&self, txid: &transaction::Hash) -> bool {
+        if let Some(evicted_at) = self.unique_entries.get(txid) {
+            // Since the list is pruned only in mutable functions, make sure
+            // we take expired items into account.
+            if !self.has_expired(evicted_at) {
+                return true;
+            }
+        }
+        false
+    }
+
+    /// Get the size of the list.
+    //
+    // Note: if this method being mutable becomes an issue, it's possible
+    // to compute the number of expired transactions and subtract.
+    pub fn len(&mut self) -> usize {
+        self.prune_old();
+        self.unique_entries.len()
+    }
+
+    /// Clear the list.
+    #[allow(dead_code)]
+    pub fn clear(&mut self) {
+        self.unique_entries.clear();
+        self.ordered_entries.clear();
+    }
+
+    /// Prune TXIDs that are older than `eviction_time` ago.
+    ///
+    // This method is public because ZIP-401 states about pruning:
+    // > This MAY be done periodically,
+    pub fn prune_old(&mut self) {
+        while let Some(txid) = self.front() {
+            let evicted_at = self
+                .unique_entries
+                .get(txid)
+                .expect("all entries should exist in both ordered_entries and unique_entries");
```

Comment:
Diagnostics Nitpick: Show the actual value.

Unfortunately, there isn't a nicer API for this.

```suggestion
                .unwrap_or_else(|_| panic!("all entries should exist in both ordered_entries and unique_entries, missing {:?} in unique_entries", txid));
```